src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs

```diff
@@ -0,0 +1,3129 @@
+                                }
+                                writer.WriteLine($"runtextpos += {c};");
+                            }
+                        }
+                        break;
+
+                    case RegexCode.Onerep | RegexCode.Rtl:
+                    case RegexCode.Notonerep | RegexCode.Rtl:
+                    case RegexCode.Setrep | RegexCode.Rtl:
+                    case RegexCode.Onerep | RegexCode.Ci | RegexCode.Rtl:
+                    case RegexCode.Notonerep | RegexCode.Ci | RegexCode.Rtl:
+                    case RegexCode.Setrep | RegexCode.Ci | RegexCode.Rtl:
+                        {
+                            int c = Operand(1);
+                            if (c != 0)
+                            {
+                                const string Length = "tmp1";
+
+                                using (EmitBlock(writer, $"if ({c} > runtextpos - runtextbeg)"))
+                                {
+                                    writer.WriteLine($"goto {Backtrack};");
+                                }
+                                writer.WriteLine($"runtextpos -= {c};");
+                                writer.WriteLine($"{Length} = {c}; // length");
+
+                                string l1 = DefineLabel();
+                                MarkLabel(l1);
+
+                                string expr = $"runtext[runtextpos + --{Length}]";
+                                if (Code() == RegexCode.Setrep)
+                                {
+                                    EmitTimeoutCheck(writer, hasTimeout);
+                                    using (EmitBlock(writer, $"if (!{MatchCharacterClass(hasTextInfo, options, expr, rm.Code.Strings[Operand(0)], IsCaseInsensitive())})"))
+                                    {
+                                        writer.WriteLine($"goto {Backtrack};");
+                                    }
+                                }
+                                else
+                                {
+                                    expr = ToLowerIfNeeded(hasTextInfo, options, expr, IsCaseInsensitive() && RegexCharClass.ParticipatesInCaseConversion(Operand(0)));
+                                    string op = Code() == RegexCode.Onerep ? "!=" : "==";
+                                    using (EmitBlock(writer, $"if ({expr} {op} {Literal((char)Operand(0))})"))
+                                    {
+                                        writer.WriteLine($"goto {Backtrack};");
+                                    }
+                                }
+
+                                using (EmitBlock(writer, $"if ({Length} > 0)"))
+                                {
+                                    writer.WriteLine($"goto {l1};");
+                                }
+                            }
+                            break;
+                        }
+
+                    case RegexCode.Oneloop:
```

Comment:
is there some clever way with pattern matching to avoid all this repetition?

Reply:
Suggestions welcome.